From 142ca29f8f32583e97d540914ed566cb50afd1d5 Mon Sep 17 00:00:00 2001
From: Lars Karlslund <8929787+lkarlslund@users.noreply.github.com>
Date: Wed, 29 Jul 2026 15:49:36 +0200
Subject: [PATCH 2145/2165] drm/msm/dp: use direct SDP flush for PSR

Newer DP controller revisions normally use the SDE peripheral-update flush
mode.  With PSR, that path can report successful frame-capture, update and
exit interrupts while the sink captures a black frame.

Select the direct UPDATE_SDP flush mode when configuring PSR.  This is also
the mode already used by older DP controller revisions.

Signed-off-by: Lars Karlslund <8929787+lkarlslund@users.noreply.github.com>
(cherry picked from commit b69a404af95952308e1adab5d20234fe9bffb5b3)

diff --git a/drivers/gpu/drm/msm/dp/dp_ctrl.c b/drivers/gpu/drm/msm/dp/dp_ctrl.c
index 9024abdc2c569f4dac7c67f8eb2e291ae778847a..cd4a73269716d1eed7cd93e9d595ec8fd54b6207 100644
--- a/drivers/gpu/drm/msm/dp/dp_ctrl.c
+++ b/drivers/gpu/drm/msm/dp/dp_ctrl.c
@@ -1855,6 +1855,15 @@ void msm_dp_ctrl_config_psr(struct msm_dp_ctrl *msm_dp_ctrl)
 	if (!ctrl->panel->psr_cap.version)
 		return;
 
+	/*
+	 * PSR SDPs must be flushed directly. The peripheral flush path can
+	 * complete normally while the sink captures an invalid frame.
+	 */
+	cfg = msm_dp_read_link(ctrl, REG_DP_MAINLINK_CTRL);
+	cfg &= ~DP_MAINLINK_CTRL_FLUSH_MODE_MASK;
+	cfg |= DP_MAINLINK_FLUSH_MODE_UPDATE_SDP;
+	msm_dp_write_link(ctrl, REG_DP_MAINLINK_CTRL, cfg);
+
 	/* enable PSR1 function */
 	cfg = msm_dp_read_link(ctrl, REG_PSR_CONFIG);
 	cfg |= PSR1_SUPPORTED;
